### PR TITLE
Added codeAction (extract subSchema to defs)

### PR DESCRIPTION
**Pull request by arpitkuriyal**
#132 
Point to Note in this:-
1. **Child Nodes of $defs Are Added from the Top**
     * Instead of adding child elements from the bottom, they are inserted from the top.
     * This prevents inconsistencies caused by offset + textLength changes due to spaces and brackets

https://github.com/user-attachments/assets/d217bc32-229a-4bbd-ad1d-c29dba4a075d

### CODE REVIEW

**Review by jdesrosiers — CHANGES_REQUESTED** [review]
This is great progress! But, it looks like there's still some work to do.

Adding the definitions to the top isn't good enough. Also, the sloppy formatting of the generated code isn't acceptable either. The biggest motivating factor for this project is to encourage best practices and good style. Two of those rules are that `$schema` and `$id` always come first and that `$defs` always goes last. It's important that the definitions are in the right place and reasonably formatted when moved. You might have to somehow run the moved code through a formatter once it's in its new location. [jsonc-parser](https://www.npmjs.com/package/jsonc-parser) is used internally to parse the schema. It has a formatting feature that you might be able to use somehow.

I noticed that the schema you're using in your demo video has a problem. The "shipping_address" property has a property "address" that then has the actual schema in it. That subschema will do nothing at all. "address" is not JSON Schema keyword, so it gets ignored. You can tell something's wrong because the schema in "address" doesn't have the expected syntax highlighting. JSON Schema keywords should be highlighted differently than plain JSON properties.

I see you're using the property name as the definition name. That's a nice default, but that's not always going to work. Subschemas can be in quite a few places other than `properties` values. For example, this would work for the `items` keyword. I think what needs to happen is that the user needs to provide the definition name. Have a look at the `if`/`then` completion provider. There's a special syntax that allows you to set placeholders that the user will be prompted to fill in. That approach may also work here.

You'll need to add tests for this feature that cover whatever edge cases you can think of.

I appreciate the clean and well written code so far. I suggest installing the ESLint plugin in your editor to get linting feedback while developing. It will help avoid getting build errors in your PRs if you forget to run the linter before pushing.

**Comment by arpitkuriyal** [comment]
Thank you for the detailed feedback. However, I encountered an issue while looking into what you mentioned:

> I think what needs to happen is that the user needs to provide the definition name.
There's a special syntax that allows you to set placeholders that the user will be prompted to fill in. That approach may also work here.

I found that `InsertTextFormat: InsertTextFormat.Snippet` is coming in LSP 3.18 ([see here](https://github.com/microsoft/language-server-protocol/issues/724)). I’ll try using the command feature as an alternative and see how it goes.

**Comment by jdesrosiers** [comment]
Thanks for figuring that out. Yes, it looks like SnippetTextEdit is what we need. Although 3.18 isn't released yet, it looks like `vscode-languageserver-node` [already supports it](https://github.com/microsoft/vscode-languageserver-node/pull/1343). That means that vscode's language server client probably supports it too. Other clients may not support it yet, but that's ok.

**Comment by arpitkuriyal** [comment]
I think it's not supported by` vscode-languageserver-node` yet.

As you can see in the screenshots, the latest available version is still 3.17.5, and this feature is introduced in 3.18. I also checked the `node_modules` directory and couldn't find any trace of this feature there.
Could you please confirm this on your end as well?

version ScreenShot
Screenshot 2025-02-22 at 1 34 48 AM

nodeModule Screenshot
Screenshot 2025-02-22 at 1 20 05 AM
PR Screenshot that u mention here:-
> Although 3.18 isn't released yet, it looks like `vscode-languageserver-node` [already supports it](https://github.com/microsoft/vscode-languageserver-node/pull/1343).
Screenshot 2025-02-22 at 1 21 54 AM

**Comment by jdesrosiers** [comment]
Yes, you're right. Although the code was merged over a year ago, it hasn't made it to an official release yet. It's planned for the next major release (v10) and that only has pre-releases published so far. We could change the dependency to `"vscode-languageserver": "^10.0.0-next" to use the pre-release. But, that's only the server. It probably wouldn't work in vscode yet.

I guess that means we have to hold off on that detail. I was hoping this would give us a way to avoid having to come up with a more robust way to generate definition names, but it looks like we're going to have to find a temporary solution until this feature is released.

**Comment by arpitkuriyal** [comment]
I think, for now, we should keep it simple and implement a basic defCounter that generates definition names like def1, def2, and so on.

Do you have any other solutions in mind? Please let me know your thoughts.

**Comment by jdesrosiers** [comment]
Number based naming can work, but there are some some edge cases.

Extract a schema and you'd get `def1`. Then restart the language server and extract another schema and you get `def1` again. Ideally people will rename after extraction and this won't come up often, but it could be a problem.

Also, if you do refactorings in one schema and then go to another schema, it would be weird for it to generate `def5` or something when there isn't `def1` - `def4` yet.

One thing that I think would work is the generate number-based names starting at `def1` and check if the name already exists. If it does, increment and check again until you have a unique name.

Or, you could inspect all the definition names looking for the `def{number}` pattern and increment starting from the number you find or 1 if the pattern isn't found.

Or, you could generate a UUID and use that as the name and not have to check anything. But, a bulky UUID might not make for as good a user experience.

Any of those options would be ok.

**Comment by arpitkuriyal** [comment]
Alright, I’ll work on it and let you know once it’s done.

**Comment by arpitkuriyal** [comment]
https://github.com/user-attachments/assets/9b392b86-4712-4c10-97e1-e07a0a987721

Please review it. If everything looks good, I will start writing the test cases.

**Comment by arpitkuriyal** [comment]
When switching the dialect URI of the schema, we need to change `$defs` to `definitions` for older drafts. Therefore, I added a condition: if the dialect is` 2020-12` or` 2019-09`, use `$defs`; otherwise, use `definitions`. Is this correct, or is there anything else I should do?

**Comment by jdesrosiers** [comment]
> Is this correct, or is there anything else I should do?

Use the `getKeywordName` function from `@hyperjump/json-schema/experimental`. It takes a keyword URI and a dialect URI and returns the right label for the dialect.

**Comment by arpitkuriyal** [comment]
Got it! I'll use` getKeywordName` from `@hyperjump/json-schema/experimental` for the right label.

**Comment by arpitkuriyal** [comment]
I just made the change, you can check it now. I'll write the test case as soon as possible.

File: language-server/src/features/codeAction/extractSubschema.js [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+        codeActionProvider: true
+      }
+    }));
+
+    // Helper function to format new def using jsonc-parser
+    const formatNewDef = (/** @type {string} */ newDefText) => {
+      try {
+        /** @type {unknown} */
+        const parsedDef = jsoncParser.parse(newDefText);
+        return JSON.stringify(parsedDef, null, 2).replace(/\n/g, "\n    ");
```
Comment (jdesrosiers) [inline language-server/src/features/codeAction/extractSubschema.js:33]: Hardcoding the indentation strategy to two spaces isn't going to work. You're going to need to determine what indentation strategy the client is using and match that. You should be able to get the information from the [configuration service](https://github.com/hyperjump-io/json-schema-language-tools/blob/main/language-server/src/services/configuration.js), but it's currently only configured to retrieve this server's configs. I think there's another "section" you'd have to request, but you'll have to figure out what that is.

File: language-server/src/features/codeAction/extractSubschema.js [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+      let highestDefNumber = 0;
+      if (definitionsNode) {
+        const defsContent = schemaDocument.textDocument.getText().slice(
+          definitionsNode.offset,
+          definitionsNode.offset + definitionsNode.textLength
+        );
+        const defMatches = [...defsContent.matchAll(/"def(\d+)":/g)];
+        defMatches.forEach((match) =>
+          highestDefNumber = Math.max(highestDefNumber, parseInt(match[1], 10))
+        );
```
Comment (jdesrosiers) [inline language-server/src/features/codeAction/extractSubschema.js:67]: This isn't a good approach. Use `keys` from `schema-node.js` to loop over all the property names of the `definitionsNode`. Then you can use the regex on those values to determine the `highestDefNumber`.

File: language-server/src/features/codeAction/extractSubschema.js [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+      }
+      let newDefName = `def${highestDefNumber + 1}`;
+      const extractedDef = schemaDocument.textDocument.getText(range);
+      const newFormattedDef = formatNewDef(extractedDef);
+      let defName = getKeywordName(
+      /** @type {string} */ (node.root.dialectUri),
```
Comment (jdesrosiers) [inline language-server/src/features/codeAction/extractSubschema.js:73]: Code style

```suggestion
        /** @type {string} */ (node.root.dialectUri),
```

File: language-server/src/features/codeAction/extractSubschema.js [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        "https://json-schema.org/keyword/definitions"
+      );
+
+      /** @type {CodeAction} */
+      const codeAction = {
+        title: `Extract '${newDefName}' to $defs`,
```
Comment (jdesrosiers) [inline language-server/src/features/codeAction/extractSubschema.js:79]: I don't think using the definition name in the title makes sense when the name is generated. Also, it should use the `defName` instead of hadcoding `$defs`

```suggestion
        title: `Extract subschema to ${defName}`,
```

**Review by jdesrosiers — CHANGES_REQUESTED** [review]
Have another try at the JSON formatting part. The other things I mentioned should be small and easy to address. I included a few code style suggestions that aren't caught by the linter. In general, I thought the code made better use of whitespace the last time I reviewed. Things are more compact now making the code harder to read.

File: language-server/src/features/codeAction/extractSubschema.js (hunk 1 above)
Comment (arpitkuriyal) [reply language-server/src/features/codeAction/extractSubschema.js:33]: I have made all the changes, but I am stuck on the formatting part. It always takes the default tab size of four instead of the current tab size. Even though I am fetching the editor settings, it doesn't seem to reflect the actual tab size used in the document. Do you have any suggestions on how to correctly retrieve the active document's indentation settings? 
Screenshot 2025-02-26 at 1 38 00 AM

Comment (jdesrosiers) [reply language-server/src/features/codeAction/extractSubschema.js:33]: Try sending the document's URI (`schemaDocument.textDocument.uri`) in the `scopeUri` parameter in the [configuration request](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_configuration). That should return the settings for that file, instead of the settings for the workspace. That's my best guess.

I think it should be ok to add an optional parameter to the `get` function so you can pass in the document URI.

Comment (arpitkuriyal) [reply language-server/src/features/codeAction/extractSubschema.js:33]: I also tried this approach, but when I console it in the terminal, it still always shows 4. Not sure why it's not picking up the actual tab size. Any other suggestions?

Comment (jdesrosiers) [reply language-server/src/features/codeAction/extractSubschema.js:37]: I'm don't have any more guesses. I'll try to find some time tonight to try some things and see if I can figure anything out.

Comment (arpitkuriyal) [reply language-server/src/features/codeAction/extractSubschema.js:37]: Okay, then I'll set detectIndentation: true and handle the indentation detection. Thanks for your help!

**Comment by arpitkuriyal** [comment]
I have used a library to detect the indentation of the current file correctly, and now everything is working fine. I also added this condition `(range.start.line === range.end.line && range.start.character === range.end.character)` in the codeAction to ensure that the extract option only appears when a subschema is selected. Previously, the bulb symbol would pop up even when we were just inside a subschema.

Additionally, in the `formatNewDef` function, I used:
`jsoncParser.applyEdits(newDefText, edits).replace(/\n/g, \n${" ".repeat(detectedIndent?.size ?? defaultTabSize)})` since I couldn't find a better option to match the overall indentation.

if put only this` jsoncParser.applyEdits(newDefText, edits)`

https://github.com/user-attachments/assets/8e1548a5-ed63-43ab-8c57-38207d66a107

with this `jsoncParser.applyEdits(newDefText, edits).replace(/\n/g, \n${" ".repeat(detectedIndent?.size ?? defaultTabSize)})`

https://github.com/user-attachments/assets/6971d22a-af13-411d-bc82-ee0169c2bd89

If you have any better ideas, please share.

**Comment by jdesrosiers** [comment]
Your problem is that you're formatting the extracted definition in isolation. The idea is to format the definition within the context of the full JSON document. That context is how it knows how much to indent.

The the edits the `format` function returns applies to the whole document, but we can adjust the offsets so they can be applied to just the new definition. Here's an example.

```javascript
import { applyEdits, format } from "jsonc-parser";

/**
 * @import { FormattingOptions, Edit } from "jsonc-parser"
 */


/** @type (text: string, edit: Edit, options: FormattingOptions) => Edit */
const withFormatting = (text, edit, options) => {
  const newText = applyEdits(text, [edit]);

  const range = { offset: edit.offset, length: edit.content.length };
  const formatEdits = format(newText, range, options);

  // Adjust the offsets so they can be applied to just the definition
  for (const formatEdit of formatEdits) {
    formatEdit.offset -= edit.offset;
  }

  return { ...edit, content: applyEdits(edit.content, formatEdits) };
};

const schema = `{
  "type": "object",
  "properties": {
    "foo": {
      "type": "number"
    },
    "bar": { "$ref": "#/$defs/bar" }
  },
  "$defs": {
    "bar": {
      "type": "boolean"
    }
  }
}`;

const offset = 51;
const length = 30;
const definition = schema.substring(offset, offset + length);

const formattingOptions = {
  keepLines: true,
  insertSpaces: true,
  tabSize: 2,
  eol: "\n"
};

/** @type Edit[] */
const edits = [
  {
    offset: 51,
    length: 30,
    content: `{ "$ref": "#/$defs/def1" }`
  },
  withFormatting(schema, {
    offset: 137,
    length: 0,
    content: `\n"def1": ${definition},`
  }, formattingOptions)
];
console.log(edits);

console.log(applyEdits(schema, edits));
```

You should be able to do something similar, but your version of the `withFormatting` function would take and return edits in the form the code action response expects.

**Comment by arpitkuriyal** [comment]
Thanks for the insight!

**Comment by jdesrosiers** [comment]
> we could modify it to find the nearest definition node and place the new definition property there.

The definition should go at the root of the schema resource it's in. A schema resource is a schema with `$id` or the root schema. The `root` on SchemaNode points to the schema resource, not necessarily the root of the document. So, I don't think there's anything special that needs to be done to get the right behavior. It should just work.

**Comment by arpitkuriyal** [comment]
now it is working fine with all the correct indentation.

> simple schema with indenation 2 spaces

https://github.com/user-attachments/assets/c5537dca-1d96-451a-878f-14c601b0a1e9

> simple schema with indentation of 4 spaces

https://github.com/user-attachments/assets/85703848-069f-49ed-863e-0ffe36af63f8

>now we can see it can correct it indentation according to the schema.

https://github.com/user-attachments/assets/435e6adb-91a8-4e61-aefd-bd1e3fadf385

**Review by jdesrosiers — CHANGES_REQUESTED** [review]
Please remove `.DS_Store`. It should be in `.gitignore` if you would like to add it.

The formatting code is sloppy. Please try to clean that up. For example, there's no reason to pass `textDocument`. You only use it to get the `uri` which you use to read the file, but you already have the text, so there's no need to read the file. It's all more complicated than it needs to be and needs more thought.

File: language-server/src/util/util.js [hunk 5]
```diff
@@ -108,3 +112,57 @@ export const readDirRecursive = async function* (path, filter, cwd) {
...
+* defaultTabSize: number, insertSpaces: boolean, detectIndentation: boolean, offset: number) => Promise<TextEdit>
+*/
+export const withFormatting = async (textDocument, text, textEdit, defaultTabSize, insertSpaces, detectIndentation, offset) => {
+  const detectedIndent = await detectIndentationFromContent(textDocument.uri, defaultTabSize, insertSpaces, detectIndentation);
+
+  const formattingOptions = {
+    insertSpaces: detectedIndent?.type === "spaces",
+    tabSize: detectedIndent?.size ?? defaultTabSize,
+    keepLines: true,
+    eol: "\n"
```
[the remark this answers is not in the document]
Comment (arpitkuriyal) [reply language-server/src/util/util.js:146]: for this we have to request another section as eol is not in the editor Settings.

**Comment by arpitkuriyal** [comment]
I have made the changes. Could you please check?

**Comment by jdesrosiers** [comment]
I did some clean up of the formatting code. Have a look.

File: language-server/src/services/configuration.js [hunk 6]
```diff
@@ -81,16 +85,30 @@ export class Configuration {
     });
   }
 
-  /** @type () => Promise<DocumentSettings> */
-  async get() {
+  /** @type (documentUri?: string) => Promise<DocumentSettings> */
+  async get(documentUri) {
```
Comment (jdesrosiers) [inline language-server/src/services/configuration.js:81]: Didn't we find that the `scopeUri` doesn't solve anything? Using `detect-indent` was what solved the problem. I think that means we don't need to pass this argument, right?

File: language-server/src/services/configuration.js [hunk 7]
```diff
@@ -81,16 +85,30 @@ export class Configuration {
...
+        { section: "jsonSchemaLanguageServer" },
+        { section: "editor", scopeUri: documentUri },
+        { section: "files.eol" }
+      ]);
+      const [extensionSettings, editorSettings, eol] = /** @type [{ defaultDialect?: string; schemaFilePatterns?: string[] },{ tabSize?: number; insertSpaces?: boolean; detectIndentation?: boolean }, string ] */ (config);
+      /** @type {{ tabSize?: number; insertSpaces?: boolean; detectIndentation?: boolean, endOfLine?: string }} */
+      const indentationSettings = {
+        tabSize: editorSettings?.tabSize ?? this.#defaultSettings.tabSize,
+        insertSpaces: editorSettings?.insertSpaces ?? this.#defaultSettings.insertSpaces,
+        detectIndentation: editorSettings?.detectIndentation ?? this.#defaultSettings.detectIndentation,
```
Comment (jdesrosiers) [inline language-server/src/services/configuration.js:94]: You're referencing `this.#defaultSettings`, but you never set a value in the default settings, so this can never do anything. I think the only one that needs a default is `detectIndentation = true`.

File: language-server/src/services/configuration.js [hunk 8]
```diff
@@ -81,16 +85,30 @@ export class Configuration {
...
-      /** @type unknown */
-      const fullSettings = { ...this.#defaultSettings, ...settings };
+      /** @type {unknown[]} */
+      const config = await this.#server.workspace.getConfiguration([
+        { section: "jsonSchemaLanguageServer" },
+        { section: "editor", scopeUri: documentUri },
+        { section: "files.eol" }
+      ]);
+      const [extensionSettings, editorSettings, eol] = /** @type [{ defaultDialect?: string; schemaFilePatterns?: string[] },{ tabSize?: number; insertSpaces?: boolean; detectIndentation?: boolean }, string ] */ (config);
+      /** @type {{ tabSize?: number; insertSpaces?: boolean; detectIndentation?: boolean, endOfLine?: string }} */
```
Comment (jdesrosiers) [inline language-server/src/services/configuration.js:90]: This is too messy. Use `@typedef` to give names to the object types that you can use here.

**Review by jdesrosiers — CHANGES_REQUESTED** [review]
Let's have another look at the Configuration changes and try to clean things up. I noticed there was some duplication in that configs were being built in two places. I handled that problem, so you don't have to worry about it, but let's see if we can make the rest a little nicer.

File: language-server/src/services/configuration.js (hunk 7 above)
Comment (arpitkuriyal) [reply language-server/src/services/configuration.js:94]: Oh, I see. I completely misunderstood this while adding it, I'll update the changes accordingly.

**Comment by arpitkuriyal** [comment]
That's a lot of learning from you! I can clearly see the difference between my approach and yours in the `withFormatting function`. Thanks for that!
i'll update you with the changes in 2-3 days as i have my college exams going on.

File: language-server/src/services/configuration.js (hunk 6 above)
Comment (arpitkuriyal) [reply language-server/src/services/configuration.js:81]: Yes now we don't need this, i'll remove it. Thanks

File: language-server/src/services/configuration.js [hunk 9]
```diff
@@ -84,13 +87,27 @@ export class Configuration {
...
+        { section: "jsonSchemaLanguageServer" },
+        { section: "editor" },
+        { section: "files.eol" }
+      ]);
+      const [extensionSettings, editorSettings, eol] = /** @type [DocumentSettings, IndentationSettings, string] */ (config);
+      /** @type IndentationSettings */
+      const indentationSettings = {
+        tabSize: editorSettings?.tabSize,
+        insertSpaces: editorSettings?.insertSpaces,
+        detectIndentation: editorSettings?.detectIndentation ?? this.#defaultSettings.detectIndentation,
```
Comment (jdesrosiers) [inline language-server/src/services/configuration.js:101]: You still haven't assigned a default setting for `detectIndentation`. See line 55.

File: language-server/src/util/util.js [hunk 10]
```diff
@@ -108,3 +113,40 @@ export const readDirRecursive = async function* (path, filter, cwd) {
...
+  const indentation = settings.detectIndentation ? detectIndent(textDocument.getText()) : {
+    amount: settings.tabSize,
+    type: settings.insertSpaces ? "space" : "tab"
+  };
+
+  const formattingOptions = {
+    insertSpaces: indentation.type === "space",
+    tabSize: indentation.amount,
+    keepLines: true,
+    eol: settings.endOfLine
```
Comment (jdesrosiers) [inline language-server/src/util/util.js:128]: I noticed that the `files.eol` setting can be "auto", which is not a valid value here. Node has `os.EOL` that we can use if `files.eol` is "auto".

File: language-server/src/services/configuration.js [hunk 11]
```diff
@@ -20,11 +21,21 @@ import ignore from "ignore";
  * }} Settings
  */
```
Comment (jdesrosiers) [inline language-server/src/services/configuration.js:22]: This type isn't being used anymore.

File: language-server/src/util/util.js [hunk 12]
```diff
@@ -108,3 +113,40 @@ export const readDirRecursive = async function* (path, filter, cwd) {
     }
   }
 };
+
+/** @type (textDocument: TextDocument, textEdit: TextEdit, settings: IndentationSettings) => TextEdit */
```
Comment (jdesrosiers) [inline language-server/src/util/util.js:117]:
```suggestion
/** @type (textDocument: TextDocument, textEdit: TextEdit, settings: DocumentSettings) => TextEdit */
```

I'm surprised we're not getting a type error here. What you're passing is the result of calling `Configuration.get`, which is `DocumentSettings`. Since you didn't update `DocumentSettings` with the all the indentation settings, I would have expected it to tell you that the value passed is not compatible with `IndentationSettings`. I think it slips through the cracks just because none of the `IndentationSettings` are required.

**Review by jdesrosiers — CHANGES_REQUESTED** [review]
There are a few issues with the types.

File: language-server/src/services/configuration.js (hunk 9 above)
Comment (arpitkuriyal) [reply language-server/src/services/configuration.js:101]: Sorry for this i totally misunderstood it. I originally thought that we needed to explicitly define the value in `settings.json` for it to pick up from the `default settings`. I don't know why i was thinking that way, but now I realize that `defaultSettings` should already include a default value for detectIndentation.

File: language-server/src/services/configuration.js [hunk 13]
```diff
@@ -11,6 +11,7 @@ import ignore from "ignore";
  * @typedef {{
  *   defaultDialect?: string;
  *   schemaFilePatterns: string[];
+ *   detectIndentation: boolean;
  * }} DocumentSettings
```
[the remark this answers is not in the document]
Comment (arpitkuriyal) [reply language-server/src/services/configuration.js:15]: should we do this here `const [extensionSettings, editorSettings, eol] = /** @type [Partial<DocumentSettings> | null, Partial<DocumentSettings> | null, string | null] */ (config);` as `DocumentSettings` have all the settings returned from the `get` function.

**Review by jdesrosiers — CHANGES_REQUESTED** [review]
It's looking much better now! I pushed a little more clean up for the Configuration class and I updated the TestClient. It originally only supported the one configuration scope. Now it can support others. You'll need this when you write tests.

One more change I'd like to see is that I want the definition to get added to the bottom of `$defs`, not the top. Then go ahead with writing tests.

Thanks for sticking with this. I told you when you took this on that it could be a tough one. The hardest part is often writing the tests, so good luck as you move into that phase.

**Comment by arpitkuriyal** [comment]
> One more change I'd like to see is that I want the definition to get added to the bottom of `$defs`, not the top

Okay i'll do it, 

> Thanks for sticking with this. I told you when you took this on that it could be a tough one

Thanks for your guidance. It’s been a challenging task, but also a great learning experience for me!

**Comment by arpitkuriyal** [comment]
> One more change I'd like to see is that I want the definition to get added to the bottom of `$defs`, not the top.

The easiest approach I found is using `definitionsNode?.children.at(-1)` to get the last definition. Alternatively, we can use `SchemaNode.keys()` to iterate over the definition node and retrieve the last definition.  Which approach do you think is better or you have something better in your mind

**Comment by jdesrosiers** [comment]
`children.at(-1)` is perfect. There's no reason to iterate. We just want the last one.

**Comment by arpitkuriyal** [comment]
I wanted to point out an edge case, if the `definitionsNode` exists but is empty (`defs: {}`), the insertion position should default to `definitionsNode.offset + 1`. Otherwise, we should find the last definition and insert the new one after it. 

https://github.com/user-attachments/assets/8684ff29-438d-4151-8c22-1038fdc75931

**Comment by jdesrosiers** [comment]
That's definitely something we'll want to have a test for.

**Comment by arpitkuriyal** [comment]
Just for review, I have written 2–3 test cases. Can you please review them and tell me if I am going in the right direction? Each test case is getting quite long. Should we write them this way only, or should I use a different approach? Can you help me with this?

File: language-server/src/features/codeAction/extractSubschema.test.ts [hunk 14]
```diff
@@ -0,0 +1,217 @@
...
+              uri: documentUri,
+              version: null
+            }
+          }
+        ]
+      },
+      kind: "refactor.extract",
+      title: "Extract 'def2' to definitions"
+    };
+    expect(codeActions?.[0]).toMatchObject(expectedCodeAction);
```
Comment (jdesrosiers) [inline language-server/src/features/codeAction/extractSubschema.test.ts:168]: Although Jest style matchers are supported by vitest, I've been sticking to chai matchers in this project. I think it's nice to keep that consistency whenever possible.

```suggestion
    expect(codeActions?.[0]).to.eql(expectedCodeAction);
```

File: language-server/src/features/codeAction/extractSubschema.test.ts [hunk 15]
```diff
@@ -0,0 +1,217 @@
...
+
+    const codeActions = await client.sendRequest(CodeActionRequest.type, {
+      textDocument: { uri: documentUri },
+      range: {
+        start: { line: 4, character: 13 },
+        end: { line: 7, character: 5 }
+      },
+      context: { diagnostics: [] }
+    });
+    expect(codeActions?.[0].title).toBe("Extract 'def1' to $defs");
```
Comment (jdesrosiers) [inline language-server/src/features/codeAction/extractSubschema.test.ts:192]: Another Jest style matcher.

```suggestion
    expect(codeActions?.[0].title).to.equal("Extract 'def1' to $defs");
```

Comment (jdesrosiers) [inline language-server/src/features/codeAction/extractSubschema.test.ts:192]: Excellent! This is a perfect example of when you don't need to check the whole response. All you're testing is that the code action was triggered. :+1:

**Review by jdesrosiers — CHANGES_REQUESTED** [review]
You're definitely on the right track. Go ahead and write more tests.

**Review by jdesrosiers — CHANGES_REQUESTED** [review]
Great work! I have one minor clean up I'd like you to look at and there are a couple more tests I'd like to see. Then we should be done!

There are a few places where the "end" comes before "start". Have a look over it again and make sure those are all ordered the way that makes sense.

I think there are a couple more tests that we should have around the configuration options. This screenshot from the test coverage report shows what's missing.

image

**Review by jdesrosiers — APPROVED** [review]
:tada:
